nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/CountText.java

```diff
@@ -0,0 +1,318 @@
+                REL_FAILURE)));
+    }
+
+    private volatile boolean countLines;
+    private volatile boolean countLinesNonEmpty;
+    private volatile boolean countWords;
+    private volatile boolean countCharacters;
+    private volatile boolean splitWordsOnSymbols;
+    private volatile String characterEncoding;
+
+    private volatile int lineCount;
+    private volatile int lineNonEmptyCount;
+    private volatile int wordCount;
+    private volatile int characterCount;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @OnScheduled
+    public void onSchedule(ProcessContext context) {
+        this.countLines = context.getProperty(TEXT_LINE_COUNT_PD).isSet()
+                ? context.getProperty(TEXT_LINE_COUNT_PD).asBoolean() : false;
+        this.countLinesNonEmpty = context.getProperty(TEXT_LINE_NONEMPTY_COUNT_PD).isSet()
+                ? context.getProperty(TEXT_LINE_NONEMPTY_COUNT_PD).asBoolean() : false;
+        this.countWords = context.getProperty(TEXT_WORD_COUNT_PD).isSet()
+                ? context.getProperty(TEXT_WORD_COUNT_PD).asBoolean() : false;
+        this.countCharacters = context.getProperty(TEXT_CHARACTER_COUNT_PD).isSet()
+                ? context.getProperty(TEXT_CHARACTER_COUNT_PD).asBoolean() : false;
+        this.splitWordsOnSymbols = context.getProperty(SPLIT_WORDS_ON_SYMBOLS_PD).isSet()
+                ? context.getProperty(SPLIT_WORDS_ON_SYMBOLS_PD).asBoolean() : false;
+        this.characterEncoding = context.getProperty(CHARACTER_ENCODING_PD).getValue();
+    }
+
+    /**
+     * Will count text attributes of the incoming stream.
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession processSession) throws ProcessException {
+        FlowFile sourceFlowFile = processSession.get();
+        if (sourceFlowFile == null) {
+            return;
+        }
+        AtomicBoolean error = new AtomicBoolean();
+
+        lineCount = 0;
+        lineNonEmptyCount = 0;
+        wordCount = 0;
+        characterCount = 0;
+
+        processSession.read(sourceFlowFile, in -> {
+            long start = System.nanoTime();
+
+            // Iterate over the lines in the text input
+            try {
+                BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(in, characterEncoding));
+                String line;
+                while ((line = bufferedReader.readLine()) != null) {
+                    if (countLines) {
+                        lineCount++;
+                    }
+
+                    if (countLinesNonEmpty) {
+                        if (line.trim().length() > 0) {
+                            lineNonEmptyCount++;
+                        }
+                    }
+
+                    if (countWords) {
+                        wordCount += countWordsInLine(line, splitWordsOnSymbols);
+                    }
+
+                    if (countCharacters) {
```

Comment:
I'm not sure that this logic is exactly right... the documentation indicates that the character count includes white space. However, in this case, it is including white space that is not a newline (excludes \r and \n that are consumed by the Buffered Reader).

Reply:
I mirrored the behavior of standard word processors which count whitespace but do not count newline and carriage returns. I'll make this clearer in the documentation.